### ISSUE TITLE
Mixed Content

### ISSUE THREAD

**Issue by Henni** [erming/shout#545]
Thumbnails and links may load images over http.
Because of this the browser marks the connection as not completely secure (mixed content).
Chrome:
![image](https://cloud.githubusercontent.com/assets/1267827/11300252/a6d058f4-8f8d-11e5-802f-2a2fc7a0eefe.png)
Firefox:
![image](https://cloud.githubusercontent.com/assets/1267827/11300236/95db97fc-8f8d-11e5-828d-529ac2daea02.png)

It would be great if shout would use https for all requests if it was configured to use https.
Thumbnails and link previews should be only loaded over https in this case.

**Comment by floogulinc** [comment]
This is not possible with the current link embedding system, as Shout simply embeds the image from its original link. So if the posted link is http, then the resulting image will use http. The text in the link previews is simply transmitted to the client by the Shout server and added to the DOM and for images, it sends the link and an `<img>` tag is added with the src attribute set to the link.

**Comment by Henni** [comment]
Shout could proxy those images to prevent the client from sending direct requests.
But this is probably another issue (even though this would solve this problem).

The first approach in my opinion would be to convert all urls to https, and if this results in a 404 simply don't show any image at all.

**Comment by floogulinc** [comment]
The proxy idea is what I was thinking would work the best but a side effect
of it would be increased bandwidth usage between client and server.

**Comment by astorije** [comment]
Indeed, mixed content is an issue indeed. However, the whole prefetching mechanism needs a lot of work and should be put in a separate plugin (help needed on a plugin mechanism ^^).

I'm going to tag this as a security issue, however I am not sure what we can do about it in the short-term.

@JocelynDelalande, any feedback on this?

**Comment by JocelynDelalande** [comment]
@astorije yes, proxying is better for both mixed-content and privacy IMHO, I don't find that increased bandwidth cost is a strong argument against this, related to benefits.

**Comment by floogulinc** [comment]
@JocelynDelalande I agree, bandwidth probably isn't too much of an issue, especially if we make sure the client side toggle prevents loading the images at all.

**Comment by astorije** [comment]
Proxying seems like a nice idea indeed. I am not too concerned with bandwidth on the server, but adds its own set of issues that design must take into account:

- Storage: The media folder will increase in size as time passes. Do we keep all images for an unlimited period of time? Chances are, pictures that are older than a couple weeks will never show up again so it can be removed safely. On the other hand, if someone scrolls back, deleted media will not show up anymore. We could accept the fact that old medias are removed and just move on with our lives, or triggers a re-proxying of the image when scrolling back, keeping it for 2 more weeks. Both solutions are acceptable to me, one is nicer (the second one), but more work.

- Security: Well, an easy way to DDoS a server. Right now, we have that memory issue, but as soon when Shout crashes, the memory gets freed so it doesn't blow up the entire server. If we store all images on the server, I could see someone dropping dozens of links of dozens of megabytes, killing the bandwidth (which becomes an issue), and filling up the disk for good (persistent after a restart). For that we can set up a max size for the folder (old images are thrown out) and a timeout for the downloads.

- Duplicates: If I send 10 times the same link to a 1MB image on a server with 10 users on that same Shout server, there will be 100 files of 1MB on the server (assuming we make their names unique). It's a problem... We could name them after their md5 hashes, so files would be overwritten instead of added, solving the storage issue, but that would still downloads the 100 files...

Any design ideas or other concerns? :)
Really, I'd much rather work/see work on a good plugin system before that though (heeeelp neeeeded!). This prefetching must be a very interesting project, but it would be much nicer in its own plugin IMO.

**Comment by xPaw** [comment]
Github itself proxies images using this: https://github.com/atmos/camo
